jiant/tasks/lm.py

```diff
@@ -172,3 +177,121 @@ def __init__(self, path, *args, **kw):
             "val": os.path.join(path, "valid.sentences.txt"),
             "test": os.path.join(path, "test.sentences.txt"),
         }
+
+
+class MaskedLanguageModelingTask(LanguageModelingTask):
+    """
+       Generic Masked Language Modeling Task
+    """
+
+    pass
+
+
+@register_task("mlm", rel_path="WikiText103/")
+class MLMTask(MaskedLanguageModelingTask):
+    """
+    Masked language modeling task on Toronto Books dataset
+    Attributes:
+        max_seq_len: (int) maximum sequence length
+        min_seq_len: (int) minimum sequence length
+        files_by_split: (dict) files for three data split (train, val, test)
+    """
+
+    def __init__(self, path, *args, **kw):
+        super().__init__(path, *args, **kw)
+        self.files_by_split = {
+            "train": os.path.join(path, "train.sentences.txt"),
+            "val": os.path.join(path, "valid.sentences.txt"),
+            "test": os.path.join(path, "test.sentences.txt"),
+        }
+
+    def get_all_labels(self):
+        """
+        For MLM, the label space is the vocabulary space of the input.
+        """
+        labels = []
+        tokenizer = get_tokenizer(self._tokenizer_name)
+        vocab_size = len(tokenizer)
+        ordered_vocab = tokenizer.convert_ids_to_tokens(range(vocab_size))
+        for word in ordered_vocab:
+            labels.append(word)
+        for path in self.files_by_split:
+            for sent in self.get_data_iter(self.files_by_split[path]):
+                for tok in sent:
+                    if tok not in labels:
+                        labels.append(tok)
+        return labels
+
+    def update_metrics(self, out, batch=None):
+        # self.scorer1(logits,labels)
+        self.scorer1(out["loss"].mean())
+        return
+
+    def get_data_iter(self, path):
+        """Loading data file and tokenizing the text
+        Args:
+            path: (str) data file path
+        """
+        import csv
+
+        f = open(path, "r")
+        reader = csv.reader(f)
+        text = list(reader)
+        moses_tokenizer = get_tokenizer("MosesTokenizer")
```

Comment:
Why do we need this?

Reply:
The wikipedia corpus available is preprocessed using MosesTokenizer, so we need to detokenize it before tokenizing it with our tokenizer of choice.